### PR TITLE
serialize mixed data

### PR DESCRIPTION
**Pull request by danivek**
Proposal of an implementation of mixed data serialization #20 

@sartian @mattii if you have time, any feedback would be appreciated.

Thanks !

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -357,6 +357,28 @@ relationships: {
 }
 ```
 
+## Serialize mixed data (dynamic type)
+
+If your data contains one or multiple object of varying type, It's possible to define a configuration object instead of the type-string as the first argument of ```serialize``` and ```serializeAsync``` function with these options :
```
Comment (mattii) [inline README.md]: Typos:
- one or multiple object**s** of varying type**s**
- **i**t's possible (*it* lowercased)
- maybe remove the `function` in the end
- options: (no space before `:`)

File: test/unit/JSONAPISerializer.test.js [hunk 2]
```diff
@@ -689,19 +792,47 @@ describe('JSONAPISerializer', function() {
...
 
-    it('should throw an error if custom schema as not been registered', function(done) {
+    it('should throw an error if custom schema has not been registered', function(done) {
       expect(function() {
         Serializer.serialize('articles', {}, 'custom');
       }).to.throw(Error, 'No schema custom registered for articles');
       done();
     });
+
+    it('should throw an error when serialize mixed data with a bad dynamic type option', function(done) {
```
Comment (mattii) [inline test/unit/JSONAPISerializer.test.js]: typo: when *serializing*

File: test/unit/JSONAPISerializer.test.js [hunk 3]
```diff
@@ -810,19 +941,48 @@ describe('JSONAPISerializer', function() {
...
 
-    it('should throw an error if custom schema as not been registered', function(done) {
+    it('should throw an error if custom schema has not been registered', function(done) {
       expect(function() {
         Serializer.serializeAsync('articles', {}, 'custom');
       }).to.throw(Error, 'No schema custom registered for articles');
       done();
     });
+
+    it('should throw an error when serialize mixed data with a bad dynamic type option', function(done) {
```
Comment (mattii) [inline test/unit/JSONAPISerializer.test.js]: typo: when serializing

File: lib/JSONAPISerializer.js [hunk 4]
```diff
@@ -332,6 +380,46 @@ module.exports = class JSONAPISerializer {
...
+    // Array data
+    if (Array.isArray(data)) {
+      return data.map(d => this.serializeMixedData(typeOption, d, included));
+    }
+
+    // Single data
+    // Resolve type from data (can be a string or a function deriving a type-string from each data-item)
+    const type = (typeof typeOption.type === 'function') ?
+      typeOption.type(data) :
+      _.get(data, typeOption.type);
```
Comment (mattii) [inline lib/JSONAPISerializer.js]: That's really only a matter of style, but I find this formatting of ternary expressions more readable:
```javascript
const type = (typeof typeOption.type === 'function')
  ? typeOption.type(data)
  : _.get(data, typeOption.type);
```
But it's just a proposal. It really depends on the way you're used to read it ;)

Comment (danivek) [inline lib/JSONAPISerializer.js]: Yes, you're right! it's better

**Comment by mattii** [comment]
@sartian, do you think this would be sufficient for your current use-case? Maybe you could give it a try in your project?

**Comment by sartian** [comment]
Hey folks, I've been in the hospital for the last week or so. All better now. I'll try to take a look soon, scrambling with a lot of catchup now. Thanks!

**Comment by danivek** [comment]
Hey folks, I also need this feature right now. So I release it.
Thanks for all !

**Comment by mattii** [comment]
Great! So someone can test it in production 👍

**Pull request merged**